### DISCUSSION TITLE
오브젝트 위치 동기화

### DISCUSSION THREAD

**Discussion by LeeYongSu1** [naverz/zepeto-studio-kor#84] [5. Q&A]
안녕하세요
현재 박스를 미는 기능을 구현하고 있습니다.
먼저 접속한 플레이어가 월드 내에 있는 박스를 밀어서 위치를 변경했고,
후에 접속한 플레이어는 먼저 접속한 플레이어의 박스 위치를 저장해서 처음 입장시 박스 위치를 바꾸고 시작하도록 작성했습니다.

문제는 기존에 접속해 있던 플레이어가 박스를 밀면 나중에 접속한 플레이어 화면에 잘 전송이 되는데 나중에 입장한 플레이어가 박스를 밀면 박스의 위치가 변하지 않는 문제가 발생합니다.

스크립트 보시고 의견 주시면 감사하겠습니다.

제가 초보라 스크립트가 지저분해도 양해 부탁드립니다....

클라이언트 입장 시 박스 위치 초기화 및 다른 클라언트의 박스 위치를 받는 스크립트 입니다.
입장시 초기화 스크립트
![클라이언트 스타터](https://user-images.githubusercontent.com/80555875/145931163-14b0a3f0-03e7-452a-b59d-50b9fa1720e6.png)
2.
![클라이언트 스타터2](https://user-images.githubusercontent.com/80555875/145931229-54a604c3-1519-409b-8520-1beb2e51fa2c.png)

박스 위치 업데이트 스크립트
![조인박스](https://user-images.githubusercontent.com/80555875/145931556-6e649c78-be44-4fe0-b188-56b10341ee9c.png)
![업데이트박스](https://user-images.githubusercontent.com/80555875/145931579-19ccac2d-05ff-462f-a2b2-3ffaeba880ad.png)

플레이어의 충돌 처리 스크립트
![플레이어 박스 푸쉬](https://user-images.githubusercontent.com/80555875/145931667-7f0b99ec-d622-4583-a983-1902461ff507.png)

박스 오브젝트에서 서버로 위치 전송하는 스크립트
![박스 위치 전송](https://user-images.githubusercontent.com/80555875/145932046-5c82fcc1-5955-45d5-9dd0-e5cf9a34e620.png)

서버 클라이언트 통신 스크립트
![index](https://user-images.githubusercontent.com/80555875/145932072-8d263995-738b-4897-9dee-a4bd4f960de1.png)

![index-onjoin](https://user-images.githubusercontent.com/80555875/145932081-feb3ae46-5b8c-4b98-9c33-f707ffa3ebe6.png)

**Comment by moondory77-naverz** [comment]
안녕하세요.
먼저 단계적으로 문제의 원인을 좁혀보시는 것을 추천드립니다.

Room에 입장 했을 때,
1. 기존 유저 -> 입장 유저 의 StateChange는 정상적으로 Broadcast 되지만
2. 입장 유저 -> 기존 유저 의 StateChange를 정상적으로 반영하지 못하는 상황

이라 할 때,
우선 Room에 입장하는 타이밍에 기존 Room에 존재하는 유저에 대한 listener가 잘 등록되고 있는지 확인이 필요할 듯 합니다. 
listener의 등록과 갱신 여부를 로그를 삽입하여 확인해 보시는것도 좋은 방법일듯 합니다.

그럼에도 계속 어려움을 겪으실 경우, 상황의 범위를 조금만 더 좁혀주시면 더 구체적인 도움을 드릴 수 있을 듯 합니다.
감사합니다.
